src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,851 @@
+                intermediate.FeatureValues = new VBuffer<float>(values.Count, values.ToArray());
+            }
+
+            private static bool SplitOneRight(ReadOnlyMemory<char> memory, char separator, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right)
+            {
+                if (memory.IsEmpty)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+
+                int index = memory.Span.LastIndexOf(separator);
+                if (index == -1)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+
+                left = memory.Slice(0, index);
+                right = memory.Slice(index + 1, memory.Length - index - 1);
+                return true;
+            }
+        }
+
+        // This class is used by the CustomMappingTransformer that does the initial parsing of the input.
+        // The features are mapped to two fields: a vector of floats for the feature values, and a vector
+        // of ReadOnlyMemory<char> for the feature indices/names.
+        private sealed class IntermediateInput
+        {
+            public float Label;
+            public float Weight;
+            public VBuffer<ReadOnlyMemory<char>> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+            public ReadOnlyMemory<char> Comment;
+            [KeyType(ulong.MaxValue - 1)]
+            public ulong GroupId;
+        }
+
+        /// <summary>
+        /// This class is used for mapping the vector of <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> field in an <see cref="IntermediateInput"/> object
+        /// to a numeric vector of indices. This class is used in case the indices in the input file are numeric. For the case where the input file contains
+        /// feature names, instead of a <see cref="CustomMappingTransformer{TSrc, TDst}"/>, we use a <see cref="ValueToKeyMappingTransformer"/>.
+        /// </summary>
+        private sealed class Indices
+        {
+            public VBuffer<int> FeatureKeys;
+
+            public static void ParseIndicesToOneBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index > 0)
+                        editor.Values[i] = index - 1;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+
+            public static void ParseIndicesToZeroBased(IntermediateInput input, Indices output)
+            {
```

Comment:
What happens if the caller asks that the data be be parsed with a one based index, but the data has a zero in its list of keys?

Reply:
The way it is written now, the feature will be ignored. (see line 390).
Do you think we should throw?

---
In reply to: [332208505](https://github.com/dotnet/machinelearning/pull/4190#discussion_r332208505) [](ancestors = 332208505)